spritecontrol.cpp

```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+			double random = Noise::random();
+			double sign = Noise::random() < 0.5 ? -1.0 : 1.0;
+			random = random * slope + pow(random, exponent) * (1 - slope);
+			return 1.0 + sign * random * variation;		// Equation: 1 + sign(rand) * (|rand| * slope + |rand|^exp * (1 - slope)) * variation
+		};
+
+		// Converts a vector to an angle in the range (-M_PI, M_PI].
+		auto get_angle = [](const Point &vector) {
+			double x = get<X>(vector);
+			double y = get<Y>(vector);
+
+			if (x == 0 && y == 0) {
+				return 0.0;
+			}
+
+			double angle = atan2(y, x);
+
+			// i'm sure there will be absolutely nothing wrong with this
+			if (x >= 0) {
+				return angle;
+			} else if (y >= 0) {
+				return angle + M_PI;
+			} else {
+				return angle - M_PI;
+			}
+		};
+
+		// Wraps angles to be inside the range (-M_PI, M_PI]
+		auto wrap_angle = [](double angle) {
+			angle += M_PI;
+			angle -= 2 * M_PI * floor(angle / (2 * M_PI));
+			return angle - 2 * M_PI;
+		};
+
+		auto get_vector_magnitude = [](const Point &vector) {
+			double x = get<X>(vector);
+			double y = get<Y>(vector);
+
+			return sqrt(x * x + y * y);
+		};
+
+		auto normalize_vector = [get_vector_magnitude](const Point &vector) {
+			double magnitude = get_vector_magnitude(vector);
+			return magnitude != 0.0 ? vector / magnitude : vector;
+		};
+
+		if (velocity.empty() || desired_velocity.empty() || desired_separation.empty() || vision_range.empty() || blind_angle.empty()) {
+			for (const Sprite *sprite : *sprites) {
+				double magnitude = Noise::random() + 0.4;
+				double radians = Noise::random() * M_PI * 2;
+				velocity[sprite->id()] = Point(std::cos(radians) * magnitude, std::sin(radians) * magnitude);
+				desired_velocity[sprite->id()] = magnitude;
+
+				double separation_radius = random_curve(0.2) * SEPARATION_X_RADIUS;
+				desired_separation[sprite->id()] = separation_radius;
+
+				double vision_radius = random_curve(0.5) * VISION_X_RADIUS;
+				vision_range[sprite->id()] = vision_radius;
+
+				double degrees = random_curve(1.0, 10.0, 0.25) * BLIND_DEGREES;
+				blind_angle[sprite->id()] = degrees * M_PI / 180.0;
+			}
+		}
+
+		std::map<Sprite *, Point> separation_velocity_changes;
+		std::map<Sprite *, Point> alignment_velocity_changes;
+		std::map<Sprite *, Point> cohesion_velocity_changes;
+		std::map<Sprite *, Point> edge_push_velocity_changes;
+
+		for (Sprite *current_sprite : *sprites) {
+			Point current_final = Point(current_sprite->final<X>(), current_sprite->final<Y>());
+			size_t sprites_seen = 1;
+			double average_angle = get_angle(velocity[current_sprite->id()]);
+			Point average_pos = Point(current_final);
+
+			for (Sprite *other_sprite : *sprites) {		// dejil... i am sorry...
+				if (current_sprite == other_sprite) {
+					continue;
+				}
+
+				Point other_final = Point(other_sprite->final<X>(), other_sprite->final<Y>());
+				Point diff = current_final - other_final;
+				get<Y>(diff) *= STRETCH_RATIO;
```

Comment:
I absolutely agree.
And additionally, the fact that it goes from -1 to 1 instead of 0 to 1 massively fucked with me earlier cause I couldn't figure out why the fuck sprites on the bottom half and the left half of the screen were getting absolutely launched.